File: 1186-maximum-subarray-sum-with-one-deletion/1186-maximum-subarray-sum-with-one-deletion.cpp
class Solution {
public:
    int maximumSum(vector<int>& arr) {
        int n = arr.size();
        int ends[n];
        int starts[n];
        for(int i=0, s=0; i < n; ++i) {
            s = max(s,0) + arr[i];
            ends[i] = s;
        }
        for(int i=n-1, s=0; ~i; --i) {
            s = max(s,0) + arr[i];
            starts[i] = s;
        }
        int ans = *max_element(ends, ends+n);
        for(int i=1; i<n-1; ++i) {
            ans = max(ans, ends[i-1] + starts[i+1]);
        }
        return ans;
    }
};

File: 1186-maximum-subarray-sum-with-one-deletion/1186-maximum-subarray-sum-with-one-deletion.py
class Solution:
    def maximumSum(self, arr: List[int]) -> int:
        n = len(arr)
        ends = [0] * n
        starts = [0] * n
        
        # Calculate the maximum subarray sum ending at each index
        s = 0
        for i in range(n):
            s = max(s, 0) + arr[i]
            ends[i] = s
        
        # Calculate the maximum subarray sum starting at each index
        s = 0
        for i in range(n - 1, -1, -1):
            s = max(s, 0) + arr[i]
            starts[i] = s
        
        # Initialize the answer with the maximum sum of a subarray without any deletions
        ans = max(ends)
        
        # Consider the maximum sum we can get by deleting one element
        for i in range(1, n - 1):
            ans = max(ans, ends[i - 1] + starts[i + 1])
        
        return ans